src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.Core.cs

```diff
@@ -244,6 +249,31 @@ internal sealed class CompilerAnalyzerAssemblyResolver(AssemblyLoadContext compi
                 }
             }
         }
+
+        private sealed class DiskResolver : IAnalyzerAssemblyResolver
+        {
+            public static readonly DiskResolver Instance = new DiskResolver();
+            public Assembly? Resolve(AnalyzerAssemblyLoader loader, AssemblyName assemblyName, AssemblyLoadContext directoryContext, string directory)
+            {
+                var realPath = loader.GetAssemblyLoadPath(assemblyName, directory);
+                return realPath is not null ? directoryContext.LoadFromAssemblyPath(realPath) : null;
+            }
+        }
+
+        private sealed class StreamResolver : IAnalyzerAssemblyResolver
+        {
+            public static readonly StreamResolver Instance = new StreamResolver();
+            public Assembly? Resolve(AnalyzerAssemblyLoader loader, AssemblyName assemblyName, AssemblyLoadContext directoryContext, string directory)
+            {
+                var realPath = loader.GetAssemblyLoadPath(assemblyName, directory);
+                if (realPath is null)
+                {
+                    return null;
+                }
+                using var stream = File.Open(realPath, FileMode.Open, FileAccess.Read, FileShare.Read);
```

Comment:
nit: might be good to also doc our IO exception story for all these types (if not already doc'ed on IAnalyzerAssemblyResolver).  namely that they will bubble up (as opposed to returning null in that case).